R-package/src/xgboost_R.cc

```diff
@@ -138,21 +138,37 @@ SEXP SafeMkChar(const char *c_str, SEXP continuation_token) {
 }
 }  // namespace
 
+char cpp_ex_msg[256];
+struct RRNGStateController {
+  RRNGStateController() {
+    GetRNGstate();
+  }
+
+  ~RRNGStateController() {
+    PutRNGstate();
+  }
+};
+
 /*!
  * \brief macro to annotate begin of api
  */
 #define R_API_BEGIN()                           \
-  GetRNGstate();                                \
-  try {
+  try {                                         \
+    RRNGStateController rng_controller{};
 /*!
  * \brief macro to annotate end of api
  */
 #define R_API_END()                             \
   } catch(dmlc::Error& e) {                     \
-    PutRNGstate();                              \
-    error(e.what());                            \
+    Rf_error("%s", e.what());                   \
+  } catch(std::exception &e) {                  \
+    std::strncpy(cpp_ex_msg, e.what(), 256);    \
+    goto throw_cpp_ex_as_R_err;                 \
   }                                             \
-  PutRNGstate();
+  if (false) {                                  \
+    throw_cpp_ex_as_R_err:                      \
+    Rf_error("%s", cpp_ex_msg);                 \
+  }
```

Comment:
is this for debugging?

Reply:
No, it's to show the C++ exception message as an R error.

Reply:
To clarify: when an exception is thrown, it gets to that condition through a `goto` statement. Would be good to hear other ideas if you think there's a better way to do this - I myself couldn't think of any other way.

Reply:
Would be great if leave some comments for explaining these few lines of code, it's not very intuitive to see some code being meaningful inside a condition block that's not reachable under normal c control flow.

Reply:
Added a comment above.